### ISSUE TITLE
Simulated Temperature Module running on child device unable to send messages upstream (to parent)

### ISSUE THREAD

**Issue by jailohani** [Azure/iotedge#6935]
**Description**
I have EFLOW 1.4 running on child and parent devices in a nested hierarchy where only the parent device has access to the internet and the child device pulls its module images from the parent.
I am following this [tutorial](https://learn.microsoft.com/en-us/azure/iot-edge/tutorial-nested-iot-edge-for-linux-on-windows?view=iotedge-1.4) and trying to run Microsoft's Simulated Temperature Module on the child device and send its telemetry messages upstream to the parent device and eventually to the IoT Hub.

The simulated temperature module is successfully started up on the child device but is unable to send telemetry data upstream to the parent. The module log output shows that it keeps waiting on trying to establish module client.

**To Reproduce**
1. Deploy EFLOW VM and setup Edge runtime on parent and child devices
2. Configure child device to pull module images from parent.
3. Verify the edge modules are running on both parent and child
4. See log output of SimulatedTemperatureSensor and edgeHub modules to verify that telemetry is not being sent to parent.

**Expected behavior**
Telemetry messages being generated and sent to parent device should be visible in module logs.

**Windows Host OS (please complete the following information):**
 - Windows Server 2019 Standard
 -  Version- 10.0.17763 Build 17763
 - Virtual Machine: Local VM
 

<details>
<summary>
Output of iotedge logs simulatedTemperatureSensor
</summary>
<p>
[2023-02-28 04:17:32 : Starting Module

SimulatedTemperatureSensor Main() started.

Initializing simulated temperature sensor to send 5 messages, at an interval of 5 seconds.

To change this, set the environment variable MessageCount to the number of messages that should be sent (set it to -1 to send unlimited messages).

[Information]: Trying to initialize module client using transport type [Amqp_Tcp_Only].
</p>
</details>

<details>
<summary>
Output of iotedge logs edgeHub
</summary>
<p>

<6> 2023-02-28 04:43:02.561 +00:00 [INF] - Attempting to connect to IoT Hub for client test-lab-edge-device-pin/$edgeHub via AMQP...
<6> 2023-02-28 04:43:02.837 +00:00 [INF] - Obtained edge hub config from module twin
<6> 2023-02-28 04:43:02.927 +00:00 [INF] - Set the following 1 route(s) in edge hub
<6> 2023-02-28 04:43:02.927 +00:00 [INF] - MetricsRoute: FROM /* INTO $upstream
<6> 2023-02-28 04:43:02.927 +00:00 [INF] - Updated message store TTL to 7200 seconds
<6> 2023-02-28 04:43:02.927 +00:00 [INF] - Updated the edge hub store and forward configuration
<6> 2023-02-28 04:43:02.927 +00:00 [INF] - Successfully completed periodic operation Get EdgeHub config
<6> 2023-02-28 04:43:50.320 +00:00 [INF] - Entering periodic task to reauthenticate connected clients
<4> 2023-02-28 04:44:02.571 +00:00 [WRN] - Error creating cloud connection for client test-lab-edge-device-pin/$edgeHub
System.TimeoutException: The operation timed out.
 ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Transient network error occurred, please retry.
 ---> System.TimeoutException: The operation has timed out.
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpConnectionHolder.EnsureConnectionAsync(TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpConnectionHolder.OpenSessionAsync(DeviceIdentity deviceIdentity, TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpUnit.EnsureSessionIsOpenAsync(TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpUnit.OpenAsync(TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.OpenAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ProtocolRoutingDelegatingHandler.OpenAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass27_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass38_0.<<OpenInternalAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.EnsureOpenedAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.OpenAsync()
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.InternalClient.OpenAsync()
   at Microsoft.Azure.Devices.Edge.Util.TaskEx.TimeoutAfter(Task task, TimeSpan timeout, Action action) in /mnt/vss/_work/1/s/edge-util/src/Microsoft.Azure.Devices.Edge.Util/TaskEx.cs:line 160
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ModuleClientWrapper.OpenAsync() in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ModuleClientWrapper.cs:line 65
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.<>c__DisplayClass30_0.<<InvokeFunc>b__0>d.MoveNext() in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 196
--- End of stack trace from previous location ---
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 189
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 189
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.OpenAsync() in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 65
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection.ConnectToIoTHub(ITokenProvider newTokenProvider) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnection.cs:line 140
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection.CreateNewCloudProxyAsync(ITokenProvider newTokenProvider) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnection.cs:line 123
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection.Create(IIdentity identity, Action`2 connectionStatusChangedHandler, ITransportSettings[] transportSettings, IMessageConverterProvider messageConverterProvider, IClientProvider clientProvider, ICloudListener cloudListener, ITokenProvider tokenProvider, TimeSpan idleTimeout, Boolean closeOnIdleTimeout, TimeSpan operationTimeout, TimeSpan cloudConnectionHangingTimeout, String productInfo, Option`1 modelId) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnection.cs:line 104
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnectionProvider.TryCreateCloudConnectionFromServiceIdentity(IIdentity identity, Action`2 connectionStatusChangedHandler, Boolean refreshOutOfDateCache, CloudListener cloudListener, String authChain) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnectionProvider.cs:line 306
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnectionProvider.ConnectInternalWithDeviceStateTracking(IIdentity identity, Action`2 connectionStatusChangedHandler, Boolean refreshCachedIdentity) in /mnt/vss/_work/1/s/edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnectionProvider.cs:line 267

</p>
</details>

**Comment by nyanzebra** [comment]
@jailohani would you mind confirming that from the child device that you can reach the parent device?

1. ping the parent device from child device
2. try ssl as well, for example `openssl s_client -connect parent_host:8883`

This should help us determine what connectivity issues you might be experiencing :)

**Comment by jailohani** [comment]
HI @nyanzebra, ping and SSL connect are working both ways. (parent->child as well as child->parent).

**Comment by nyanzebra** [comment]
> HI @nyanzebra, ping and SSL connect are working both ways. (parent->child as well as child->parent).

Interesting, that means device to device connection is fine... what about from eflow vm and from within container? If it works from eflow vm and not from container (can try `docker exec --it <my pod> /bin/sh`) then it might be there is a parent_hostname typo in config.toml or a resolution issue between container and vm. If it fails in vm, then similar issue but need to make sure vm gets same network settings as host.

Sorry to ask you to try twice more, but my current guess on likely culprit is somewhere at the vm or container parent_host:8883 either doesn't resolve or something. I will see if can try to replicate locally as well.

**Comment by jailohani** [comment]
The connection between parent EFLOW VM and Child EFLOW VM is fine as well (ping and ssl is working). That made it possible for the child EFLOW VM to pull its module images from the parent (upstream:443). The iotedge check command shows that container on the default network can connect upstream to the HTTPS, AMQP and MQTT ports successfully.

I'll check ping and ssl from within the container also and let you know in sometime (right now our VM is down for some hardware maintenance issues).

Thanks.

**Comment by jlian** [comment]
> I'll check ping and ssl from within the container also and let you know in sometime (right now our VM is down for some hardware maintenance issues).

Any updates?

**Comment by jailohani** [comment]
Hi, our hardware is still down. Will update as soon as its back online.
Thanks

**Comment by jailohani** [comment]
Hi @jlian,
We had deployed EFLOW VM on our parent and child with the Default configuration i.e. 1GB of RAM.

We redeployed EFLOW with 2GB and 4GB of RAM on child and parent respectively and we were able to get the module working! (Child is able to send telemetry to IoTHub via the Parent).

So I think we can close this issue?  Just 1 question- Can we reopen this issue incase of the same issue in the future? Or would we need to create a new issue? 

Thanks for your help!
